bin/dada2_denoising.R

```diff
@@ -0,0 +1,136 @@
+library(dada2)
+library(tidyverse)
+library(argparse)
+
+parser <- ArgumentParser(description='DADA2 Denoising')
+parser$add_argument('--derep-1', type="character", required=TRUE, nargs="+", help="Path to RDS file containing dereplicated forward sequences")
+parser$add_argument('--derep-2', type="character", required=TRUE, nargs="+", help="Path to RDS file containing dereplicated reverse sequences")
+parser$add_argument('--error-model-1', type="character", required=TRUE, help="Path to the forward error model RDS file")
+parser$add_argument('--error-model-2', type="character", required=TRUE, help="Path to the reverse error model RDS file")
+parser$add_argument('--ncores', type="numeric", required=TRUE, help="Number of threads to use")
+parser$add_argument('--pool', type="character", default="false")
+parser$add_argument('--band-size', type='integer', default=16)
+parser$add_argument('--omega-a', type='double', default=1e-120)
+parser$add_argument('--just-concatenate', action='store_true')
+parser$add_argument('--use-quals', type="character", default="false")
+parser$add_argument('--maxEE', type="integer", default=2)
+parser$add_argument('--self-consist', action='store_true')
+parser$add_argument('--omega-c', type='double', default=1e-40)
+parser$add_argument('--ampliconFILE', type="character", required=TRUE, help="Path to amplicon info file")
+parser$add_argument('--verbose', action="store_true", help="Verbose")
+
+args <- parser$parse_args()
+print(args)
+
+# Load error models
+err_model_F <- readRDS(args$error_model_1)
+err_model_R <- readRDS(args$error_model_2)
+
+# Function to gather derep objects from a list of files
+gather_dereps <- function(paths) {
+    all_dereps <- list()
+    for(path in paths) {
+        derep <- readRDS(path)
+        # Extract the sample name using a regex pattern
+        sample_name <- sub("(.*)_trimmed.*", "\\1", basename(path))
+        all_dereps[[sample_name]] <- derep
+    }
+    all_dereps
+}
+
+# Gather all derep objects
+derepFs <- gather_dereps(args$derep_1)
+derepRs <- gather_dereps(args$derep_2)
+
+# DADA2 denoising
+dadaFs <- dada(derepFs, err=err_model_F, multithread=args$ncores)
+dadaRs <- dada(derepRs, err=err_model_R, multithread=args$ncores)
+
+# Save the DADA2 denoised data
+save(dadaFs, file="dada_denoised_F.RData")
+save(dadaRs, file="dada_denoised_R.RData")
+
+# Filter samples based on the provided condition
+filter_samples <- function(samples, condition) {
```

Comment:
small nit -- I prefer putting all function declarations at the top of the script, after the argument parsing

Reply:
- [ ] functions at top